File: FindAllAnagram.java
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : None

public class FindAllAnagram {
    class Solution {
        public List<Integer> findAnagrams(String s, String p) {
            int sLen = s.length(), pLen = p.length();
            List<Integer> result = new ArrayList<>();

            if(pLen > sLen) return result;

            Map<Character, Integer> map = new HashMap<>();
            for(int i = 0; i < pLen; i++){
                char c = p.charAt(i);
                map.put(c, map.getOrDefault(c, 0) + 1);
            }

            int match = 0;
            int start = 0;
            for(int i = 0; i < sLen; i++){
                char in = s.charAt(i);

                if(map.containsKey(in)){
                    int count = map.get(in);
                    count--;
                    map.put(in, count);
                    if(count == 0)
                        match++;
                }

                if(i >= pLen){
                    char out = s.charAt(start);
                    if(map.containsKey(out)){
                        int count = map.get(out);
                        count++;
                        map.put(out, count);
                        if(count == 1)
                            match--;
                    }
                    start++;
                }

                if(match == map.size()){
                    result.add(start);
                }
            }
            return result;
        }
    }
}

File: find_all_anagram.py
# Time Complexity : Add - O(2n)
# Space Complexity :O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No
'''
1. Sliding window using slow and fast pointers

2. Maintain 2 counter array of length 26 for a string and pattern.
3. The counter will store the fequencies of any letter indexed at ascii value
4. Now we move acrros the string, updating string counter, when the length of window == len(pattern)
5. We check if strin counter array == patterm=n counter array, if yes, append left index of window in result
6. We now move left by1, and right by one, but before that we undo the changes by decrementing count value
   at repective ascii 
'''

class Solution:
       
    def compare_arrays(self, arr1, arr2):
        
        for i in range(26):
            if arr1[i] != arr2[i]:
                return False
            
        return True
        
        
    def findAnagrams(self, s: str, p: str) -> List[int]:
        
        if s == "":
            return 
        
        string_count = [0]*26
        pattern_count = [0]*26
        
        for ch in p:
            pattern_count[ord(ch)-ord('a')] += 1
            
            
        result = []
        left, right = 0, 0
        
        while right < len(s):
            
            string_count[ord(s[right]) - ord('a')] += 1
            
            if right-left+1 == len(p):
                #compare array                
                if self.compare_arrays(string_count, pattern_count):
                    result.append(left)
                
                string_count[ord(s[left]) - ord('a')] -= 1
                left +=1  
            
            right += 1
        return result
        